### ISSUE TITLE
Long form target inference is not properly versioned

### ISSUE THREAD

**Issue by emillon** [ocaml/dune#3230]
Hi,

Since #2494 we can use target inference for long-form actions, ie things like this work without an explicit `(targets)`:

    (rule
     (action
      (with-stdout-to x.txt
       (echo works))))

However, this feature does not have a `since` check, so it can be used in a `(lang dune 1.2)` project (it should be available only in >= 2.0).

**Comment by rgrinberg** [comment]
This might be a WONTFIX since addressing this issue has the potential to break even more code.

**Comment by emillon** [comment]
Yeah, the fact that it's on a major language boundary means that our usual strategy (add a nonfatal warning) results in something not very actionable (upgrading the project to dune 2.0). Backporting to a new lang 1.x would work but I don't think we want to go this route.
